src/widgets/query-rule-custom-data/query-rule-custom-data.tsx

```diff
@@ -0,0 +1,79 @@
+import React, { render, unmountComponentAtNode } from 'preact-compat';
+import {
+  getContainerNode,
+  prepareTemplateProps,
+  createDocumentationMessageGenerator,
+} from '../../lib/utils';
+import { WidgetFactory } from '../../types';
+import connectQueryRules, {
+  QueryRulesRenderer,
+} from '../../connectors/query-rules/connectQueryRules';
+import CustomData from '../../components/QueryRuleCustomData/QueryRuleCustomData';
+
+export type QueryRuleCustomDataTemplates = {
+  default: string | ((items: object[]) => string);
+};
+
+type QueryRuleCustomDataWidgetParams = {
+  container: string | HTMLElement;
+  templates?: QueryRuleCustomDataTemplates;
+  transformItems?: (items: object[]) => any;
+};
+
+interface QueryRuleCustomDataConnectorWidgetParams
+  extends QueryRuleCustomDataWidgetParams {
+  container: HTMLElement;
+}
+
+type QueryRuleCustomData = WidgetFactory<QueryRuleCustomDataWidgetParams>;
+
+const withUsage = createDocumentationMessageGenerator({
+  name: 'query-rule-custom-data',
+});
+
+const renderer: QueryRulesRenderer<
+  QueryRuleCustomDataConnectorWidgetParams
+> = ({ userData, instantSearchInstance, widgetParams }) => {
+  const { container, templates } = widgetParams;
+
+  const templateProps = prepareTemplateProps({
+    templates,
+    templatesConfig: instantSearchInstance.templatesConfig,
+  });
+
+  render(
+    <CustomData templateProps={templateProps} items={userData} />,
+    container
+  );
+};
+
+const queryRuleCustomData: QueryRuleCustomData = (
+  {
+    container,
+    templates: userTemplates = {},
+    transformItems = items => items,
+  } = {} as QueryRuleCustomDataWidgetParams
```

Comment:
We have to investigate why we have to cast here, but definitely not a blocker.

Reply:
In this case it's because `container` is required in `QueryRuleCustomDataWidgetParams`.

Since we want to keep the pattern of `const widget = (options = {})` and `container` doesn't exist in `{}`, it's necessary to cast `{}` to `QueryRuleCustomDataWidgetParams`. Does that make sense?

Reply:
Yes.